sqlmesh/core/engine_adapter/mssql.py

```diff
@@ -71,79 +73,36 @@ class for other postgres.
-        with self.temp_table(query_or_df, reference_table) as temp_table:
+        batch_size: int,
+        target_table: t.Optional[TableName] = None,
+    ) -> SourceQuery:
+        def generator(
+            df: pd.DataFrame,
+            columns_to_types: t.Dict[str, exp.DataType],
+            target_table: t.Optional[TableName] = None,
+        ) -> t.Generator[Query, None, None]:
+            temp_table = self._get_temp_table(target_table or "pandas")
+            self.create_table(temp_table, columns_to_types)
             rows: t.List[t.Iterable[t.Any]] = list(df.itertuples(False, None))
-
             conn = self._connection_pool.get()
             conn.bulk_copy(temp_table.name, rows)
-
-            yield exp.select(*columns_to_types).from_(temp_table)
-
-    def _insert_overwrite_by_condition(
-        self,
-        table_name: TableName,
-        query_or_df: QueryOrDF,
-        where: t.Optional[exp.Condition] = None,
-        columns_to_types: t.Optional[t.Dict[str, exp.DataType]] = None,
-    ) -> None:
-        """
-        SQL Server does not directly support `INSERT OVERWRITE` but it does
-        support `MERGE` with a `False` condition and delete that mimics an
-        `INSERT OVERWRITE`. Based on documentation this should have the same
-        runtime performance as `INSERT OVERWRITE`.
-
-        If a Pandas DataFrame is provided, it will be loaded into a temporary
-        table and then merged with the target table. This temporary table is
-        deleted after the merge is complete or after it's expiration time has
-        passed.
-        """
-        with self.__try_load_pandas_to_temp_table(
-            table_name,
-            query_or_df,
-            columns_to_types,
-        ) as source_table:
-            query = self._add_where_to_query(source_table, where)
-
-            columns = [
-                exp.to_column(col)
-                for col in (columns_to_types or [col.alias_or_name for col in query.expressions])
-            ]
-            when_not_matched_by_source = exp.When(
-                matched=False,
-                source=True,
-                condition=where,
-                then=exp.Delete(),
-            )
-            when_not_matched_by_target = exp.When(
-                matched=False,
-                source=False,
-                then=exp.Insert(
-                    this=exp.Tuple(expressions=columns),
-                    expression=exp.Tuple(expressions=columns),
-                ),
-            )
-            self._merge(
-                target_table=table_name,
-                source_table=query,
-                on=exp.condition("1=2"),
-                match_expressions=[when_not_matched_by_source, when_not_matched_by_target],
-            )
+            try:
+                yield exp.select(*columns_to_types).from_(temp_table)
+            finally:
+                self.drop_table(temp_table)
```

Comment:
why do all the callers have to drop their own tables?

Reply:
I'm not sure what we mean by caller here. The caller of the generator doesn't drop any tables. They get drop automatically after they fully consume the iterator. So effectively "behind the scenes" the table gets dropped but it is not something the caller needs to worry about.